### PR TITLE
Add Spin adapter

### LINKED ISSUE #68: Spin Adapter
Add Spin as a first-class deployment target, enabling EdgeZero apps to deploy to Fermyon Cloud and any WASI-compatible Spin runtime.

See [ROADMAP.md](https://github.com/stackpop/edgezero/blob/main/ROADMAP.md#epic-4-spin-adapter)

### LINKED ISSUE #69: As a developer I want to deploy EdgeZero apps to Spin
Applications should be deployable to Spin runtimes with the same code used for Fastly/Cloudflare/Axum

### LINKED ISSUE #70: Spin request conversion
Map Spin HTTP request to edgezero-core::Request

### LINKED ISSUE #71: Spin response conversion
Map edgezero-core::Response to Spin HTTP response

### LINKED ISSUE #72: As a developer I want to make outbound HTTP requests from Spin
Proxy client support so handlers can call upstream services when running on Spin

### LINKED ISSUE #75: As a developer I want to build and deploy Spin targets from the CLI
The EdgeZero CLI should support build and deploy for Spin the same way it does for Fastly and Cloudflare

### LINKED ISSUE #76: spin.toml generation
Generate spin.toml from edgezero.toml manifest

### LINKED ISSUE #78: Spin app-demo example
Port app-demo to Spin target as reference implementation

### PR DESCRIPTION
**Pull request by ChristianPavilonis**
## Summary

Adds a new `edgezero-adapter-spin` crate that bridges Spin HTTP components into the EdgeZero router, following the established adapter pattern (Fastly, Cloudflare).

- **New crate**: `crates/edgezero-adapter-spin/` with context, request, response, proxy, and CLI modules
- **Scaffold support**: Handlebars templates for `cargo edgezero new --adapter spin` scaffolding
- **Example adapter**: `examples/app-demo/crates/app-demo-adapter-spin/` reference implementation
- **Consistent API**: `init_logger()`, `AppExt` trait, `run_app()` convenience entry point — matching Fastly/Cloudflare adapters

## What's Included

| Module | Description |
|--------|-------------|
| `context.rs` | `SpinRequestContext` — extracts `spin-client-addr` and `spin-full-url` headers |
| `request.rs` | Converts `spin_sdk::http::IncomingRequest` → core `Request` with proper WASI resource handle management |
| `response.rs` | Converts core `Response` → `spin_sdk::http::Response` (buffered + streaming bodies) |
| `proxy.rs` | `SpinProxyClient` using `spin_sdk::http::send` for outbound HTTP |
| `cli.rs` | Build (`cargo build --target wasm32-wasip1`), deploy (`spin deploy`), serve (`spin up`) commands |
| `lib.rs` | `init_logger()`, `AppExt` trait, `run_app()` entry point |

## What's Not Included (Future Work)

- KV store support

## Testing

- `cargo fmt --all -- --check` — clean
- `cargo clippy --workspace --all-targets --all-features -- -D warnings` — clean
- `cargo test --workspace --all-targets` — all tests pass (6 Spin adapter tests including context unit tests)
- `cargo check --workspace --all-targets --features "fastly cloudflare"` — clean
- Contract tests in `tests/contract.rs` (gated behind `spin + wasm32`, follows Fastly/Cloudflare pattern)
- Deployed mocktioneer to Fermyon

### CODE REVIEW

File: crates/edgezero-adapter-spin/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+#[cfg(all(feature = "spin", target_arch = "wasm32"))]
+impl AppExt for edgezero_core::app::App {
+    fn dispatch<'a>(
+        &'a self,
+        req: spin_sdk::http::IncomingRequest,
+    ) -> ::core::pin::Pin<
+        Box<dyn ::core::future::Future<Output = anyhow::Result<spin_sdk::http::Response>> + 'a>,
+    > {
```
Comment (aram356) [inline crates/edgezero-adapter-spin/src/lib.rs]: 🔧 **`AppExt` duplicates `dispatch()` logic instead of delegating**: This body inlines the same convert → route → convert logic as `request::dispatch()`. If `dispatch()` changes, `AppExt` silently diverges. Both Cloudflare and Fastly adapters delegate to the standalone function.

**Fix**:
```rust
impl AppExt for edgezero_core::app::App {
    fn dispatch<'a>(
        &'a self,
        req: spin_sdk::http::IncomingRequest,
    ) -> ::core::pin::Pin<
        Box<dyn ::core::future::Future<Output = anyhow::Result<spin_sdk::http::Response>> + 'a>,
    > {
        Box::pin(crate::request::dispatch(self, req))
    }
}
```

File: crates/edgezero-adapter-spin/tests/contract.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        Ok(response)
+    }
+
+    let router = RouterService::builder()
+        .get("/uri", capture_uri)
+        .post("/mirror", mirror_body)
+        .build();
+
+    App::new(router)
+}
```
Comment (aram356) [inline crates/edgezero-adapter-spin/tests/contract.rs]: 🔧 **Contract test file has zero test cases**: `build_test_app()` and handlers exist but there are no `#[test]` or `#[wasm_bindgen_test]` functions. `cargo test` confirms `running 0 tests` for this file.

Fastly has 6 contract tests, Cloudflare has 6. Spin has 0.

**Fix**: Add at minimum request conversion, response conversion, and dispatch round-trip tests (matching the Fastly/Cloudflare pattern).

File: crates/edgezero-adapter-spin/src/proxy.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            .await
+            .map_err(|e| EdgeError::internal(format!("Spin outbound HTTP error: {e}")))?;
+
+        let status = StatusCode::from_u16(*spin_response.status())
+            .unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
+
+        // Collect response headers before consuming the body.
+        let response_headers: Vec<_> = spin_response
+            .headers()
+            .filter_map(|(name, value)| {
```
Comment (aram356) [inline crates/edgezero-adapter-spin/src/proxy.rs]: 🤔 **Silent header drops in proxy response**: Three `filter_map` drop points here — non-UTF-8 values (`as_str()?`), invalid header names (`from_bytes().ok()?`), invalid header values (`.parse().ok()?`) — all silently discarded with no logging. Upstream response headers could vanish with no indication.

Suggestion: Add `log::debug!` when a header can't be converted, so proxy issues are diagnosable.

File: crates/edgezero-adapter-spin/src/response.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+/// Both `Body::Once` and `Body::Stream` are converted to a buffered
+/// byte body. Streaming bodies are collected into a single `Vec<u8>`.
+pub async fn from_core_response(response: Response) -> Result<spin_http::Response, EdgeError> {
+    let (parts, body) = response.into_parts();
+
+    let mut builder = spin_http::Response::builder();
+    builder.status(parts.status.as_u16());
+
+    for (name, value) in parts.headers.iter() {
+        if let Ok(v) = value.to_str() {
```
Comment (aram356) [inline crates/edgezero-adapter-spin/src/response.rs]: 🤔 **Silent header drops in response conversion**: Same pattern as `proxy.rs` — `if let Ok(v) = value.to_str()` silently skips non-UTF-8 response headers with no logging.

**Review by aram356 — CHANGES_REQUESTED** [review]
## PR Review

### Summary

Solid adapter implementation that follows the established Fastly/Cloudflare pattern well. However, the adapter does not compile for its actual wasm32-wasip1 target, and the scaffolding pipeline produces broken projects due to missing workspace deps and unresolved template variables. These must be fixed before merge.

### 😃 Praise

- CLI scaffolding is thorough and well-tested — build/deploy/serve, blueprint registration, `ctor` auto-registration, manifest discovery with distance-based sorting, and 4 solid unit tests
- Context module is correctly ungated from `target_arch` so it works in native tests
- Method conversion (`into_core_method`) handles `Method::Other` with proper error reporting — more robust than the Fastly adapter

### Findings

#### Blocking

- 🔧 **wasm32 compile error**: `EdgeError::internal(format!(...))` — `String` doesn't impl `Into<anyhow::Error>` (`proxy.rs:48`)
- 🔧 **Scaffold `{crate_name}` placeholder not replaced**: generator only handles `{crate}` / `{crate_dir}` (`cli.rs:178`)
- 🔧 **`proj_spin_underscored` template var never populated**: renders as empty → broken `spin.toml` source path (`spin.toml.hbs:12`)
- 🔧 **Generated projects fail**: `spin-sdk` missing from `seed_workspace_dependencies()` in `generator.rs:109`
- 🔧 **`AppExt` duplicates `dispatch()` logic**: should delegate like Cloudflare/Fastly (`lib.rs:42-56`)
- 🔧 **Zero contract test cases**: `build_test_app()` exists but no `#[test]` functions (`tests/contract.rs`)
- 🔧 **CI feature check missing `spin`**: `.github/workflows/test.yml:59` only checks `fastly cloudflare`

#### Non-blocking

- 🤔 **Silent header drops in proxy.rs and response.rs**: non-UTF-8/invalid headers discarded with no logging
- 🤔 **`client_addr` stored as raw `String`**: Fastly uses typed `Option<IpAddr>` — cross-adapter inconsistency
- ♻️ **`init_logger()` identical cfg branches**: both return `Ok(())`, gate adds no value
- ♻️ **Body stream collection duplicated**: identical `Body::Stream → Vec<u8>` in proxy.rs and response.rs
- ⛏ **Template duplicate dependency**: `edgezero-adapter-spin` in both `[dependencies]` and `[target.wasm32.dependencies]`

### 📌 Out of Scope

- 🌱 **No compression/decompression in proxy**: Fastly transparently decompresses gzip/brotli upstream responses. Spin passes compressed bytes through raw. Worth tracking for parity.
- 📌 **`CLAUDE.md` not updated**: workspace layout, compilation targets, CI features all omit Spin.

### CI Status

- fmt: PASS
- clippy: PASS
- tests: PASS (native)
- feature check (`fastly cloudflare`): PASS
- **wasm32-wasip1 compile: FAIL** (`cargo check -p edgezero-adapter-spin --target wasm32-wasip1 --features spin`)

**Review by prk-Jr — COMMENTED** [review]
Can we update the documentation and readme.

**Review by aram356 — CHANGES_REQUESTED** [review]
## Follow-up review — critical pathing bugs

The previous review addressed compile errors and code-quality issues. This review focuses on **two runtime-breaking bugs** that cause `edgezero-cli build --adapter spin` to fail on any hyphenated crate name or workspace member. Both are reproducible in the `examples/app-demo` demo.

### Summary of findings

| Sev | Issue | Impact |
|-----|-------|--------|
| High | Artifact lookup uses raw package name (hyphens) but Cargo emits underscores | `build --adapter spin` exits with "compiled artifact not found" for any hyphenated crate |
| High | `spin.toml` `source` path points to crate-local `target/`, but workspace builds emit to workspace `target/` | `spin up` fails to find the wasm binary for any workspace member |
| Medium | `find_workspace_root` climbs to the topmost `Cargo.toml`, overshooting nested workspaces like `examples/app-demo/` | Build/deploy commands may search the wrong tree |
| Medium | Non-UTF-8 headers silently dropped across request/response/proxy boundaries | Breaks binary headers (signatures, auth tokens) in proxy edge cases |
| Medium | Test suite only exercises simple (non-hyphenated, flat workspace) artifact lookup — misses both regressions above | False green on CI |

File: crates/edgezero-adapter-spin/src/cli.rs [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+                .ok_or("invalid Cargo manifest path")?,
+        ])
+        .args(extra_args)
+        .status()
+        .map_err(|e| format!("failed to run cargo build: {e}"))?;
+    if !status.success() {
+        return Err(format!("cargo build failed with status {status}"));
+    }
+
+    let workspace_root = find_workspace_root(manifest_dir);
```
[the remark this answers is not in the document]
Comment (ChristianPavilonis) [reply crates/edgezero-adapter-spin/src/cli.rs]: Fixed in `cli_support.rs` — `find_workspace_root` now stops climbing when it finds a `Cargo.toml` containing `[workspace]`. Added `workspace_root_stops_at_workspace_table` test that verifies nested workspaces resolve to the inner workspace root.

**Comment by ChristianPavilonis** [comment]
## Review Fixes — Follow-up Review (Mar 11)

All findings from @aram356's follow-up review addressed in commit e9f4548.

| File | Change |
|------|--------|
| `crates/edgezero-adapter-spin/src/cli.rs` | Fix `locate_artifact` to use `crate_name.replace('-', "_")` for .wasm filename; fix `dest` copy target; add `locate_artifact_converts_hyphens_to_underscores` test |
| `crates/edgezero-adapter-fastly/src/cli.rs` | Same hyphen→underscore fix for `locate_artifact` and `dest` |
| `crates/edgezero-adapter-cloudflare/src/cli.rs` | Same hyphen→underscore fix for `locate_artifact` and `dest` |
| `crates/edgezero-adapter/src/cli_support.rs` | `find_workspace_root` now stops at `[workspace]` tables; add `workspace_root_stops_at_workspace_table` test |
| `crates/edgezero-adapter-spin/src/request.rs` | Use `HeaderValue::from_bytes(value)` to preserve non-UTF-8 inbound headers |
| `crates/edgezero-adapter-spin/src/proxy.rs` | Use `HeaderValue::from_bytes` for proxy response headers; document WASI limitation on outbound |
| `crates/edgezero-adapter-spin/src/response.rs` | Document WASI limitation for outbound non-UTF-8 headers |
| `crates/edgezero-adapter-spin/src/templates/spin.toml.hbs` | Use `{{target_dir_spin}}` for correct workspace-relative source path |
| `crates/edgezero-cli/src/generator.rs` | Compute `target_dir_{id}` template variable from crate depth |
| `examples/app-demo/.../spin.toml` | Fix source path to `../../target/...` |

### Verification

- `cargo fmt --all -- --check` ✓
- `cargo clippy --workspace --all-targets --all-features -- -D warnings` ✓
- `cargo test --workspace --all-targets` — 323 tests pass ✓
- `cargo check --workspace --all-targets --features "fastly cloudflare spin"` ✓

File: crates/edgezero-adapter-spin/src/proxy.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            match edgezero_core::http::HeaderValue::from_bytes(value.as_bytes()) {
+                Ok(hval) => response_headers.push((hname, hval)),
+                Err(_) => {
+                    log::warn!("dropping invalid proxy response header value for: {}", name);
+                }
+            }
+        }
+
+        let response_body = spin_response.into_body();
+        let mut proxy_response = ProxyResponse::new(status, Body::from(response_body));
```
Comment (aram356) [inline crates/edgezero-adapter-spin/src/proxy.rs]: 🔧 **Missing compression/decompression handling**

Both the Fastly proxy (`proxy.rs:161-180`) and Cloudflare proxy (`proxy.rs:84-116`) detect `Content-Encoding` (gzip/brotli) on upstream responses and decompress via `transform_stream()` from `edgezero_core::compression`. They also strip `Content-Encoding` and `Content-Length` headers after decompression.

The Spin proxy passes compressed bytes through verbatim. If an upstream responds with `Content-Encoding: gzip`, handlers will see garbled bytes — breaking the adapter contract.

**Fix**: After collecting response headers, check for `Content-Encoding` and apply `decode_gzip_stream`/`decode_brotli_stream`. Since Spin's body is already buffered, wrap it in a single-chunk stream for the decompression utilities, then strip the encoding headers.

File: crates/edgezero-adapter-spin/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+pub use context::SpinRequestContext;
+#[cfg(all(feature = "spin", target_arch = "wasm32"))]
+pub use proxy::SpinProxyClient;
+#[cfg(all(feature = "spin", target_arch = "wasm32"))]
+pub use request::{dispatch, into_core_request};
+#[cfg(all(feature = "spin", target_arch = "wasm32"))]
+pub use response::from_core_response;
+
+pub fn init_logger() -> Result<(), log::SetLoggerError> {
```
Comment (aram356) [inline crates/edgezero-adapter-spin/src/lib.rs]: 🔧 **`init_logger()` missing dual feature-gate definitions**

Cloudflare and Fastly both define `init_logger()` twice — one behind the platform feature gate, one behind `#[cfg(not(...))]` as a no-op fallback. This adapter has a single unconditional definition.

**Fix**:
```rust
#[cfg(all(feature = "spin", target_arch = "wasm32"))]
pub fn init_logger() -> Result<(), log::SetLoggerError> { Ok(()) }

#[cfg(not(all(feature = "spin", target_arch = "wasm32")))]
pub fn init_logger() -> Result<(), log::SetLoggerError> { Ok(()) }
```

File: crates/edgezero-adapter-spin/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+/// #[http_component]
+/// async fn handle(req: spin_sdk::http::IncomingRequest) -> anyhow::Result<impl spin_sdk::http::IntoResponse> {
+///     edgezero_adapter_spin::run_app::<App>(req).await
+/// }
+/// ```
+#[cfg(all(feature = "spin", target_arch = "wasm32"))]
+pub async fn run_app<A: edgezero_core::app::Hooks>(
+    req: spin_sdk::http::IncomingRequest,
+) -> anyhow::Result<impl spin_sdk::http::IntoResponse> {
+    init_logger().expect("init spin logger");
```
Comment (aram356) [inline crates/edgezero-adapter-spin/src/lib.rs]: ❓ **`init_logger()` called per-request — will panic with a real logger**

In Spin, `#[http_component]` is called per-request, so `init_logger().expect("init spin logger")` runs on every incoming request. Currently a no-op, but once a real logger is wired in, `log::set_logger` returns `Err` on the second call and this will panic.

Is this intentional? Consider using `try_init` semantics or ignoring the error:
```rust
let _ = init_logger();
```

**Review by aram356 — CHANGES_REQUESTED** [review]
## PR Review

### Summary

Solid Spin adapter implementation that closely follows the established Fastly/Cloudflare patterns. Two blocking issues around proxy compression handling and logger gating need to be addressed before merge.

### Findings

#### Blocking

- 🔧 **Proxy client missing compression/decompression**: The Spin proxy passes upstream compressed responses through verbatim, breaking the adapter contract (`proxy.rs:62`)
- 🔧 **`init_logger()` missing dual feature-gate definitions**: Diverges from Fastly/Cloudflare pattern; blocks future real logger integration (`lib.rs:32`)
- ❓ **`run_app` calls `init_logger()` on every request**: In Spin's per-request model, this will panic once a real logger is wired in (`lib.rs:67`)

#### Non-blocking

- 🤔 **README claims "Stable" for untested adapter**: Consider "Beta" or "Preview" until production-validated
- 🌱 **`collect_body_bytes` has no size limit**: Unbounded `Vec` growth; pre-existing cross-adapter concern
- 🏕 **Contract tests don't exercise routing on host**: Host-side tests are smoke tests only; consider dispatching through `oneshot()`
- ⛏ **Example `spin.toml` hardcoded relative artifact path**: Breaks silently if crate is moved

### CI Status

- fmt: PASS
- clippy: PASS
- tests: PASS (259 tests)

**Review by prk-Jr — CHANGES_REQUESTED** [review]
## PR Review

### Summary

Adds the Fermyon Spin adapter (`edgezero-adapter-spin`) with full request/response conversion, proxy support with synchronous decompression, CLI scaffolding, and WASM-compatible contract tests. The implementation is structurally sound and follows project conventions closely, but two blocking issues must be resolved before merge: an unbounded decompressed body allocation in the proxy that creates a zip-bomb risk, and missing scaffold test assertions for the newly unconditional spin registration.

---

### 😃 Praise

- **Synchronous decompression rationale is well-documented** (`proxy.rs:97-102`) — the comment explaining why async decompression is avoided in WASI's single-threaded executor is exactly the context future contributors need.
- **`let _ = init_logger()` with explanatory comment** (`lib.rs:77`) — correctly handles the per-request invocation model where `set_logger` errors after the first call; silent discard is the right defensive pattern here.
- **Clean cfg gating throughout** — the split between `SpinRequestContext` (always available) and WASM-gated `proxy`/`request`/`response` modules is precise and follows project conventions.
- **`HeaderValue::from_bytes` for inbound headers** (`request.rs:33-40`) — correctly handles non-UTF-8 header values with a `log::warn` on drop, mirroring the Cloudflare adapter pattern from prior review feedback.

---

### Findings

#### Blocking

- 🔧 **Zip-bomb risk**: `decompress_body()` has no output size cap — a 16 MiB gzip payload can expand to >1 GB. (`proxy.rs:103-123`)
- ❓ **Missing scaffold test assertions for spin**: `generate_new_scaffolds_workspace_layout` checks for `cloudflare` and `fastly` but not `spin` — the test would pass silently if spin scaffolding broke.
- ❓ **Inbound request body has no adapter-level size cap**: asymmetry with outbound bodies (16 MiB cap) is undocumented. (`request.rs:50-53`)

#### Non-blocking

- 🤔 **Spin dep in both `[dependencies]` and `[target.wasm32.dependencies]`**: Cargo merges features, activating the `spin` feature on native builds too. Consistent with the Cloudflare template — confirm intentional. (`Cargo.toml.hbs:17,21`)
- ♻️ **Duplicate `read_package_prefers_package_table` test**: verbatim copy already covered in `cli_support.rs` — remove. (`cli.rs:316-323`)
- 🏕 **Scaffold test incomplete now that spin is always registered**: 3 `assert!` lines would prevent future regressions.
- ⛏ **`"x-edgezero-proxy"` is a bare string literal**: repeated across adapters — a shared constant in `edgezero-core` would prevent typos. (`proxy.rs:88-90`)

---

### 📌 Out of Scope

- `locate_artifact` is structurally duplicated across all three adapter CLI modules. Should be extracted to `cli_support.rs` with a `target_triple` parameter in a follow-up.
- `brotli` and `flate2` are unconditional deps not gated behind the `spin` feature, so they compile into `edgezero-cli` even though only used in WASM-gated modules. Consistent with other adapters — worth addressing if binary size becomes a concern.

---

### CI Status

- fmt: PASS
- clippy: FAIL (pre-existing lint in `edgezero-macros`, not introduced by this PR)
- tests: PASS (329 tests)

**Review by aram356 — APPROVED** [review]
👍 Should be good to go once tests are fixed

**Review by aram356 — CHANGES_REQUESTED** [review]
@ChristianPavilonis Please resolve conflict

**Review by aram356 — APPROVED** [review]
👍 Ready to go

**Review by prk-Jr — APPROVED** [review]
LGTM
